presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/MergeAdjacentWindows.java

```diff
@@ -46,7 +47,16 @@
     @Override
     public Optional<PlanNode> apply(WindowNode parent, Captures captures, Context context)
     {
-        WindowNode child = captures.get(CHILD);
+        // Pulling the descendant WindowNode above projects is done as a part of this rule, as opposed in a
+        // separate rule, because that pullup is not useful on its own, and could be undone by other rules.
+        // For example, a rule could insert a project-off node between adjacent WindowNodes that use different
+        // input symbols.
```

Comment:
1.  This patch improves the existing window-merging feature to support LAG(X, 1) which is a use case for FINRA, for example, and is thus very useful.  For a demonstration of this, see the change to TestMergeWindows.testIdenticalWindowSpecificationsAAcpA.
2.  Separating iterative optimizer rules into different optimizers defeats a main purpose of the iterative optimizer, and that is to allow the rules to work together to discover new plans that the developer might not have anticipated.
3.  I agree that simply moving WC above P is not useful, and that's why this patch does not contain a rule that would do so.  This patch only does such a move as a part of merging or reordering two window nodes.
4.  Even in the future, when the planner can explore multiple alternatives, it will be important to help steer it to explore useful alternatives, as opposed to enumerating all possible plans and hoping for the best.  There are simply too many possible plans to explore even a small fraction of them.